src/System.Device.Gpio/System/Device/I2c/Drivers/UnixI2cDevice.Linux.cs

```diff
@@ -174,6 +174,9 @@ public override unsafe byte ReadByte()
         /// </param>
         public override unsafe void Read(Span<byte> buffer)
         {
+            if (buffer.Length == 0)
```

Comment:
>Length [](start = 23, length = 6)

why we throw? should we just return?

Reply:
this comment apply in all changed places.

---
In reply to: [259633550](https://github.com/dotnet/iot/pull/278#discussion_r259633550) [](ancestors = 259633550)

Reply:
zero length is pretty much always a bug on the caller side - why read nothing?

Reply:
I would suggest to have this method return the read count instead of void and don't throw here. return 0 for 0 length buffer is good indication what could be the issue. 

My point is, this is a public API and you never know the caller scenario which may end up passing zero length buffer. throwing exception is always a headache in such scenarios. we have similar notion like in Encoding and passing empty string, just a valid input and we don't throw. 


---
In reply to: [259636817](https://github.com/dotnet/iot/pull/278#discussion_r259636817) [](ancestors = 259636817)

Reply:
@tarekgh this is slightly different to regular Stream Reads/Writes - you have to know up front how many bytes you want to read and scenarios where you don't read the whole buffer are unexpected cc: @JeremyKuhne to double check I'm not saying anything incorrectly

Reply:
if reading less is possible then I agree we should return int - so far I haven't seen such case

Reply:
This API takes Span which I expect the callers will just do Slicing, ending up with empty span is possible. passing empty span and returning 0 is reasonable to me :-) 

> if reading less is possible then I agree we should return int - so far I haven't seen such case

What happen if I allocate 1 GB and send span with that length? does the read block till reading 1 GB?

Reply:
@tarekgh I do not believe there currently any devices which support reading like that - currently pretty much every device I worked with required writing an address just before doing any read and number of bytes is known in the specification - I don't think there are any scenarios for this kind of streaming here

Reply:
Are you saying, the span length always has to match the number of bytes we read? this doesn't sound right to me.

Reply:
@tarekgh I'm almost sure that is this case. I think considering the current shape of the API the empty read does not make sense - we can reiterate on the return value of this API later - does that sound reasonable? So far I haven't seen a case where you don't know much about what the other device will send you

Reply:
go ahead and don't block on my comment. Just make sure you track discussing this more and get some feedback if we need to change the behavior. thanks!

Reply:
created https://github.com/dotnet/iot/issues/288 to track this